src/core/net/dnssd_server.cpp

```diff
@@ -1209,6 +1219,8 @@ void Server::HandleTimer(void)
 
         if (query.GetExpireTime() <= now)
         {
+            mCounters.mHandledByUpstream.mFailedQueries.mNoResponse++;
+            mCounters.mHandledByUpstream.mFailedQueries.mTotalFailure++;
```

Comment:
Suggest moving these to `Server::OnUpstreamQueryDone()` callback. It should be called with `nullptr` response to indicate timeout from `otPlatDnsCancelUpstreamQuery()`.

Reply:
moved to Server::ResetUpstreamQueryTransaction()

Reply:
wait, I need to keep the totalfailure++ here as it's a timeout failure

Reply:
I think we may count twice with this. 
- Here in `HandleTimer()` we update the counter.
- The `otPlatDnsCancelUpstreamQuery()` call that follows will invoke `otPlatDnsUpstreamQueryDone()` 
- Which will in turn call `OnUpstreamQueryDone()` from which we call `ResetUpstreamQueryTransaction()` 
- And the counters are again updated in `ResetUpstreamQueryTransaction()`.

Better to update the counter from one place in `ResetUpstreamQueryTransaction()` based on the passed-in `error` (it is good/safe model as we update the counter when we delete/free the associated trasncation object with a upstream query)
- We may also need to update the `OnUpstreamQueryDone()` code to make sure the situation where `aResponseMessage` is `nullptr` is treated as an `error` and not success.

Reply:
Thanks for the clear guide!!! Updated!